src/libasr/pass/intrinsic_array_function_registry.h

```diff
@@ -488,7 +684,26 @@ static inline ASR::expr_t* instantiate_ArrIntrinsic(Allocator &al,
     int result_dims = extract_n_dims_from_ttype(return_type);
     ASR::expr_t* return_var = nullptr;
     if( result_dims > 0 ) {
-        fill_func_arg("result", return_type)
+        ASR::ttype_t* return_type_ = return_type;
+        if( !ASRUtils::is_fixed_size_array(return_type) ) {
+            bool is_allocatable = ASRUtils::is_allocatable(return_type);
+            Vec<ASR::dimension_t> empty_dims;
+            empty_dims.reserve(al, result_dims);
+            for( int idim = 0; idim < result_dims; idim++ ) {
+                ASR::dimension_t empty_dim;
+                empty_dim.loc = loc;
+                empty_dim.m_start = nullptr;
+                empty_dim.m_length = nullptr;
+                empty_dims.push_back(al, empty_dim);
+            }
+            return_type_ = ASRUtils::make_Array_t_util(al, loc,
+                ASRUtils::extract_type(return_type_), empty_dims.p, empty_dims.size());
+            if( is_allocatable ) {
+                return_type_ = ASRUtils::TYPE(ASR::make_Allocatable_t(al, loc, return_type_));
+            }
+        }
+        ASR::expr_t *result = declare("result", return_type_, Out);
+        args.push_back(al, result);
```

Comment:
Not relevant to sync

---

I am not sure why we need this, ideally `result` shall be `res(merge(xx, yy, zz),....)`. I think this is done to fix some bug but this may not be the best way to proceed. We can discuss about this after PR.

Reply:
> I am not sure why we need this, ideally `result` shall be `res(merge(xx, yy, zz),....)`.

This is applicable to functions. What we are creating above is a subroutine. The `intent(out)` variable of a subroutine can have array type with deferred dimensions. That's what being done here. You call merge while allocating the result variable while calling this subroutine. That's what we do in `intrinsic_functions.cpp` pass. The logic is correct here and doesn't need any modification.

Reply:
Ahh okay, thanks for clarifying :)